### ISSUE TITLE
preprocessing.DummyEncoder.transform() gives confusing error message with unexpected columns

### ISSUE THREAD

**Issue by ecurtin2** [dask/dask-ml#357]
When DummyEncoder encounters a dataframe with different columns than the fit() encountered, the error message prints `self.columns`. If no columns value was passed to __init__ this reports an error message like:

```python

ValueError: Columns of 'X' do not match the training columns. Got Index([.......], dtype='object'), expected None
```

If instead the error printed `self.columns_`, which appears to be set from the dataframe passed to fit() it would make the error much more clear. 

Personally, I've added something like this to our local codebase

```python 

expected = set(self.columns_)  
got = set(X.columns)

raise ValueError(
f"Columns of 'X' do not match the training columns.\n"
f"X is missing: {expected - got or 'No columns missing'}. "
f"X has but these were not expected: {got - expected or 'No unexpected columns'}."
)
```

Which makes it clear where the columns differ from expected. I'm more than happy to submit a PR for this, but have never contributed to open source before so this may take me more time than should be necessary for such a small error. 

I'm also unsure if the library has a way to handle a more general input validation across transformers, or if that is even within the scope of dask-ml.

**Comment by TomAugspurger** [comment]
Thanks. I think that's correct, we should be using `self.columns_` instead of `self.columns`.

If you are interested in making a fix, http://ml.dask.org/contributing.html has instructions for getting started. Otherwise, I'll try to get to it next week.

> I'm also unsure if the library has a way to handle a more general input validation across transformers, or if that is even within the scope of dask-ml.

Could you explain that a bit more? Generally, estimators just do whatever validation is required for their own correctness (checking shapes, missing data, etc.).

**Comment by ecurtin2** [comment]
Thanks for getting back! 

By validation I mean that ensuring the columns passed to transform() have the same name and dtype given during fit() (or at least compatible dtype, say int8 vs int32 or something). We recently ran into a bit of an issue where we did pd.read_sql(...) and the query returned the proper columns / names, but one of the inferred dtypes was unexpected since the column in the DB was NULL / dtype unspecified (I think it inferred to string). 

After writing I'm unsure this falls under the scope of the library, but It's something I'm probably gonna have to implement for work anyway.

**Comment by TomAugspurger** [comment]
> By validation I mean that ensuring the columns passed to transform() have the same name and dtype given during fit()

We typically follow scikit-learn here, which I believe will attempt to coerce things (so the int8 vs int32 wouldn't matter).

DummyEncoder is a bit special since it's dataframe-only.

For those looking to contribute, I think the relevant change is on https://github.com/dask/dask-ml/blob/cc4837c2c2101f9302cac38354b55754263cd1f3/dask_ml/preprocessing/data.py#L635. That and tests.

**Comment by TomAugspurger** [comment]
Closed by #610
